### PR TITLE
C++ MaxPool Module

### PR DESCRIPTION
**Pull request by ShahriarRezghi**

### CODE REVIEW

File: torch/csrc/api/include/torch/nn/modules/maxpool.h [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+  TORCH_ARG(ExpandingArray<D>, kernel_size);
+
+  /// the stride of the window. Default value is `kernel_size
+  TORCH_ARG(ExpandingArray<D>, stride);
+
+  /// implicit zero padding to be added on both sides
+  TORCH_ARG(ExpandingArray<D>, padding) = 0;
+
+  /// a parameter that controls the stride of elements in the window
+  TORCH_ARG(ExpandingArray<D>, dialation) = 1;
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/modules/maxpool.h]: nit: `dialation` -> `dilation`

File: torch/csrc/api/include/torch/nn/modules/maxpool.h [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+  /// The options with which this `Module` was constructed.
+  MaxPoolOptions<D> options;
+};
+
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ MaxPool1d ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+/// Applies maxpool over a 1-D input.
+/// See https://pytorch.org/docs/master/nn.html#torch.nn.MaxPool1d to learn
+/// about the exact behavior of this module.
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/modules/maxpool.h]: We might want to write the formulas here similar to https://pytorch.org/docs/stable/nn.html#maxpool1d (we can either do it in this PR or in a follow-up PR).

File: test/cpp/api/modules.cpp [hunk 3]
```diff
@@ -103,6 +104,59 @@ TEST_F(ModulesTest, Conv3d) {
...
 }
 
+TEST_F(ModulesTest, MaxPool1d) {
+  MaxPool1d model(MaxPool1dOptions(3).stride(2));
+  auto x = torch::randn({1, 1, 5}, torch::requires_grad());
+  auto y = model(x);
+  torch::Tensor s = y.sum();
+
+  s.backward();
+  std::cout << y.sizes() << std::endl;
```
Comment (yf225) [inline test/cpp/api/modules.cpp]: nit: need to remove print statement

File: test/cpp/api/modules.cpp [hunk 4]
```diff
@@ -103,6 +104,59 @@ TEST_F(ModulesTest, Conv3d) {
...
+  ASSERT_EQ(y.ndimension(), 3);
+  ASSERT_EQ(s.ndimension(), 0);
+  ASSERT_EQ(y.sizes(), torch::IntArrayRef({1, 1, 2}));
```
Comment (yf225) [inline test/cpp/api/modules.cpp]: Besides checking `ndimension()` and `sizes()`, can we also check that the values of `y` are what we expected? (we might need to set value of `x` in a specific way in order to test this)

File: test/cpp/api/modules.cpp [hunk 5]
```diff
@@ -103,6 +104,59 @@ TEST_F(ModulesTest, Conv3d) {
...
+TEST_F(ModulesTest, MaxPool2dEven) {
+  MaxPool2d model(MaxPool2dOptions(3).stride(2));
+  auto x = torch::randn({2, 5, 5}, torch::requires_grad());
+  auto y = model(x);
+  torch::Tensor s = y.sum();
+
+  s.backward();
+  ASSERT_EQ(y.ndimension(), 3);
+  ASSERT_EQ(s.ndimension(), 0);
+  ASSERT_EQ(y.sizes(), torch::IntArrayRef({2, 2, 2}));
```
Comment (yf225) [inline test/cpp/api/modules.cpp]: Ditto for checking value of `y`.

File: test/cpp/api/modules.cpp [hunk 6]
```diff
@@ -103,6 +104,59 @@ TEST_F(ModulesTest, Conv3d) {
...
+  auto x = torch::randn({2, 5, 4}, torch::requires_grad());
+  auto y = model(x);
+  torch::Tensor s = y.sum();
+
+  s.backward();
+  ASSERT_EQ(y.ndimension(), 3);
+  ASSERT_EQ(s.ndimension(), 0);
+  for (auto i = 0; i < 3; i++) {
+    ASSERT_EQ(y.size(i), 2);
+  }
```
Comment (yf225) [inline test/cpp/api/modules.cpp]: Ditto for checking value of `y`.

File: test/cpp/api/modules.cpp [hunk 7]
```diff
@@ -103,6 +104,59 @@ TEST_F(ModulesTest, Conv3d) {
...
+  auto x = torch::randn({2, 5, 5, 5}, torch::requires_grad());
+  auto y = model(x);
+  torch::Tensor s = y.sum();
+
+  s.backward();
+  ASSERT_EQ(y.ndimension(), 4);
+  ASSERT_EQ(s.ndimension(), 0);
+  for (auto i = 0; i < 4; i++) {
+    ASSERT_EQ(y.size(i), 2);
+  }
```
Comment (yf225) [inline test/cpp/api/modules.cpp]: Ditto for checking value of `y`.

**Review by yf225 — COMMENTED** [review]
@ShahriarSS Thanks and it looks fantastic! I left some very minor comments. We might also want to wait until https://github.com/pytorch/pytorch/pull/23852 is merged, so that `MaxPool` can be the first module that achieves Python/C++ API parity =D

**Comment by ShahriarSS** [comment]
@yf225 There were some problems and one test wasn't implemented. I fixed them so please take a look.

**Review by yf225 — APPROVED** [review]
Thanks a lot @ShahriarSS ! I will add the parity test in a follow-up PR.

File: caffe2/CMakeLists.txt [hunk 8]
```diff
@@ -515,6 +515,7 @@ if (NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
       ${TORCH_SRC_DIR}/csrc/api/src/nn/modules/embedding.cpp
       ${TORCH_SRC_DIR}/csrc/api/src/nn/modules/functional.cpp
       ${TORCH_SRC_DIR}/csrc/api/src/nn/modules/linear.cpp
+      ${TORCH_SRC_DIR}/csrc/api/src/nn/modules/maxpool.cpp
```
Comment (yf225) [inline caffe2/CMakeLists.txt]: @ShahriarSS the Facebook internal build seems to fail, and I am thinking we might need to add the `maxpool.cpp` entry into https://github.com/pytorch/pytorch/blob/13292ec3c7f47228b93dbef1ca750b66b4b8e0af/tools/build_variables.py#L178-L205

File: torch/csrc/api/src/nn/modules/maxpool.cpp [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+Tensor MaxPool1dImpl::forward(const Tensor& input) {
+  return torch::max_pool1d(
+      input,
+      options.kernel_size_,
+      options.stride_,
+      options.padding_,
+      options.dilation_,
+      options.ceil_mode_);
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/maxpool.cpp]: We might be missing `options.return_indices_` here (and the forward calls for 2d and 3d), I will push a commit to add it.

Comment (ShahriarSS) [reply torch/csrc/api/src/nn/modules/maxpool.cpp]: Thanks. But I don't think that ```torch::max_poolxd``` uses it. That's why I didn't include it.

Comment (ShahriarSS) [reply torch/csrc/api/src/nn/modules/maxpool.cpp]: Yes @yf225 here is the documentation: 
```
return_indices: if ``True``, will return the max indices along with the outputs.
                        Useful for :class:`torch.nn.MaxUnpool2d` later
```

Comment (yf225) [reply torch/csrc/api/src/nn/modules/maxpool.cpp]: Ah got it, thanks a lot for the catch!

**Comment by yf225** [comment]
cc. @glaringlee

**Pull request closed without merging**